provider/formtype_test.go

```diff
@@ -0,0 +1,426 @@
+package provider_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/coder/terraform-provider-coder/v2/provider"
+)
+
+// formTypeTestCase is the config for a single test case.
+type formTypeTestCase struct {
+	name        string
+	config      formTypeCheck
+	assert      paramAssert
+	expectError *regexp.Regexp
+}
+
+// paramAssert is asserted on the provider's parsed terraform state.
+type paramAssert struct {
+	FormType provider.ParameterFormType
+	Type     provider.OptionType
+	Styling  json.RawMessage
+}
+
+// formTypeCheck is a struct that helps build the terraform config
+type formTypeCheck struct {
+	formType   provider.ParameterFormType
+	optionType provider.OptionType
+	options    bool
+
+	// optional to inform the assert
+	customOptions []string
+	defValue      string
+	styling       json.RawMessage
+}
+
+func (c formTypeCheck) String() string {
+	return fmt.Sprintf("%s_%s_%t", c.formType, c.optionType, c.options)
+}
+
+func TestValidateFormType(t *testing.T) {
+	t.Parallel()
+
+	// formTypesChecked keeps track of all checks run. It will be used to
+	// ensure all combinations of form_type and option_type are tested.
+	// All untested options are assumed to throw an error.
+	formTypesChecked := make(map[string]struct{})
+
+	expectType := func(expected provider.ParameterFormType, opts formTypeCheck) formTypeTestCase {
+		ftname := opts.formType
+		if ftname == "" {
+			ftname = "default"
+		}
+
+		if opts.styling == nil {
+			// Try passing arbitrary data in, as anything should be accepted
+			opts.styling, _ = json.Marshal(map[string]any{
```

Comment:
👍 but we should also ensure that `nil` is accepted.

Reply:
Yup! That is the default: https://github.com/coder/terraform-provider-coder/blob/stevenmasley%2Fform_control/provider/parameter.go#L262-L262